File: 1578. Minimum Time to Make Rope Colorful.cpp
class Solution {
public:
    int minCost(string colors, vector<int>& neededTime) {
        stack<int> st;
        st.push(0);
        int ans = 0;

        for (int i = 1; i < colors.size(); i++) {
            int top = st.top();
            if (colors[top] == colors[i]) {
                if (neededTime[top] > neededTime[i]) {
                    ans += neededTime[i];
                } else {
                    ans += neededTime[top];
                    st.pop();
                    st.push(i);
                }
            } else {
                st.push(i);
            }
        }

        return ans;
    }
};

File: 1578. Minimum Time to Make Rope Colorful.java
class Solution {
    public int minCost(String colors, int[] neededTime) {
        Stack<Integer> st = new Stack<>();
        st.push(0);
        int ans = 0;
        for(int i = 1; i < colors.length(); i++){
            int top = st.peek();
            if(colors.charAt(top) == colors.charAt(i)){
                if(neededTime[top] > neededTime[i]){
                    ans += neededTime[i];
                }
                else{
                    ans += neededTime[st.pop()];
                    st.push(i);
                }
            }
            else st.push(i);
        }

        return ans;
    }
}